examples/megssm/mne_util.py

```diff
@@ -0,0 +1,290 @@
+Carray32 = lambda X: np.require(X, dtype=np.float32, requirements='C')
+Carray = Carray64
+
+
+class ROIToSourceMap(object):
+    """ class for computing ROI-to-source space mapping matrix 
+    
+    Notes
+    -----
+    The following variables defined here correspond to various matrices
+    defined in :footcite:`yang_state-space_2016`:
+    - fwd_src_snsr : G
+    - fwd_roi_snsr : C
+    - fwd_src_roi : L
+    - snsr_cov : Q_e
+    - roi_cov : Q
+    - roi_cov_0 : Q0         """
+
+    def __init__(self, fwd, labels, label_flip=False):
+
+        src = fwd['src']
+
+        roiidx = list()
+        vertidx = list()
+
+        n_lhverts = len(src[0]['vertno'])
+        n_rhverts = len(src[1]['vertno'])
+        n_verts = n_lhverts + n_rhverts
+        offsets = {'lh': 0, 'rh': n_lhverts}
+
+        hemis = {'lh': 0, 'rh': 1}
+
+        # index vector of which ROI a source point belongs to
+        which_roi = np.zeros(n_verts, dtype=np.int64)
+
+        data = []
+        for li, lab in enumerate(labels):
+
+            this_data = np.round(label_sign_flip(lab, src))
+            if not label_flip:
+                this_data.fill(1.)
+            data.append(this_data)
+            if isinstance(lab, mne.Label):
+                comp_labs = [lab]
+            elif isinstance(lab, mne.BiHemiLabel):
+                comp_labs = [lab.lh, lab.rh]
+
+            for clab in comp_labs:
+                hemi = clab.hemi
+                hi = 0 if hemi == 'lh' else 1
+
+                lverts = clab.get_vertices_used(vertices=src[hi]['vertno'])
+
+                # gets the indices in the source space vertex array, not the huge
+                # array.
+                # use `src[hi]['vertno'][lverts]` to get surface vertex indices to
+                # plot.
+                lverts = np.searchsorted(src[hi]['vertno'], lverts)
+                lverts += offsets[hemi]
+                vertidx.extend(lverts)
+                roiidx.extend(np.full(lverts.size, li, dtype=np.int64))
+
+                # add 1 b/c 0 corresponds to unassigned variance
+                which_roi[lverts] = li + 1
+
+        N = len(labels)
+        M = n_verts
+
+        # construct sparse fwd_src_roi matrix
+        data = np.concatenate(data)
+        vertidx = np.array(vertidx, int)
+        roiidx = np.array(roiidx, int)
+        assert data.shape == vertidx.shape == roiidx.shape
+        fwd_src_roi = csc_matrix((data, (vertidx, roiidx)), shape=(M, N))
+
+        self.fwd = fwd
+        self.fwd_src_roi = fwd_src_roi
+        self.which_roi = which_roi
+        self.offsets = offsets
+        self.n_lhverts = n_lhverts
+        self.n_rhverts = n_rhverts
+        self.labels = labels
+
+        return
+
+    @property
+    def fwd_src_sn(self):
+        return self.fwd['sol']['data']
+
+    @property
+    def fwd_src_roi(self):
+        return self._fwd_src_roi
+
+    @fwd_src_roi.setter
+    def fwd_src_roi(self, val):
+        self._fwd_src_roi = val
+
+    @property
```

Comment:
@jadrew43 this is line 114 and it's not related to scaling. There is some stuff below that is. Can you confirm by commenting on the correct lines/lines in the "Files" tab of this PR? This not being the correct line makes me wonder if you forgot a push, and I don't want to look prematurely.

The best way to help me check would be to make it very easy to run and check the results. For example, add a commented-out (to make CIs happy) `# np.testing.assert_allclose(data, data_mne, rtol=..., atol=...)` statement that fails (when you uncomment it) where you scale `data` the old way and `data_mne` the new way. Then I need a minimal script (hopefully runs in just a few seconds, and on `sample` data for example) I can run on this branch that will fail when I uncomment the `assert_allclose` line

Reply:
The updated files are in `state_space/`, I haven't done anything in `examples/` in weeks so I went ahead and deleted the files I had in there. Running `state_space_connectivity.py` takes about 90 seconds, adding `_mne_scale_sensor_data()` adds about 60 seconds. Happy to chat about building tests that take much less time. The commented testing line is L228 in `mne_util.py`.

Reply:
Hmm, I don't think this is quite a minimal reproducible bit of code.

When I tried to run the code, it failed because I needed `autograd`. I installed that, then it complained that I needed `autograd_linalg`. I can't find this on `pip` or `conda`, so I tried swapping in `from scipy.linalg import solve_triangular`.

Next I tried to run the script and got:
```
FileNotFoundError: [Errno 2] No such file or directory: '/home/larsoner/mne_data/MNE-sample-data/MEG/sample/labels/AUD-lh.label'
```
This makes sense, usually labels are in the SUBJECTS_DIR. But even then there aren't auditory labels in `sample`. So let's change it to something that should work I think:
```
regexp = '^(G_temp_sup-Plan_tempo.*|Pole_occipital)'
labels = mne.read_labels_from_annot(
    'sample', 'aparc.a2009s', regexp=regexp, subjects_dir=subjects_dir)
label_names = [label.name for label in labels]
assert len(label_names) == 4
```
Then I got a bit farther, until I hit:
```
AttributeError: python: undefined symbol: mkl_get_max_threads
```
So I then disabled all the thread-setting stuff, and then get to:
```
  File "/home/larsoner/python/scipy/scipy/linalg/_basic.py", line 336, in solve_triangular
    raise ValueError('expected square matrix')
ValueError: expected square matrix
```
So my swap of `scipy.linalg` is not correct.

I don't think I can proceed until I can install `autograd_linalg` somehow, how do you recommend doing this?

Reply:
Try this:
pip install --src deps -e git+ssh://git@github.com/nfoti/autograd_linalg.git@master#egg=autograd_linalg

Reply:
I pushed a commit for a `solve_triangular` that allows the code to run on my machine. I can look into the issue now!